graphql/src/store/prefetch.rs

```diff
@@ -699,101 +699,77 @@ fn collect_fields<'a>(
-                                                .or_default()
-                                                .append(&mut group);
-                                        }
-                                    }
-                                }
-                            });
+                    if visited_fragments.insert(&spread.fragment_name) {
+                        let fragment = ctx.query.get_fragment(&spread.fragment_name);
+                        let fragment_ty = {
+                            let q::TypeCondition::On(ty_name) = &fragment.type_condition;
+                            object_or_interface_by_name(schema, ty_name).ok_or_else(|| {
+                                QueryExecutionError::NamedTypeError(ty_name.clone())
+                            })?
+                        };
+
+                        if fragment_ty.is_object() {
+                            type_condition = fragment_ty
+                        };
+
+                        collect_fields_inner(
+                            ctx,
+                            fragment_ty,
+                            once(&fragment.selection_set),
+                            visited_fragments,
+                            output,
+                        )?;
                     }
                 }
 
                 q::Selection::InlineFragment(fragment) => {
-                    let fragment_cond = TypeCondition::from(fragment.type_condition.clone());
-                    // Fields for this fragment need to be looked up in the type
-                    // mentioned in the condition
-                    let fragment_type =
-                        fragment_cond.matching_type(&ctx.query.schema.document, object_type);
-
-                    // The `None` case here indicates an error where the type condition
-                    // mentions a nonexistent type; the overall query execution logic will catch
-                    // that
-                    if let Some(fragment_type) = fragment_type {
-                        let fragment_grouped_field_set = collect_fields(
-                            ctx,
-                            &fragment_type,
-                            vec![&fragment.selection_set],
-                            Some(visited_fragments.clone()),
-                        );
-
-                        for (response_key, type_fields) in fragment_grouped_field_set {
-                            for (type_cond, mut group) in type_fields {
-                                if let Some(cond) = fragment_cond.and(&type_cond) {
-                                    grouped_fields
-                                        .entry(response_key)
-                                        .or_default()
-                                        .entry(cond)
-                                        .or_default()
-                                        .append(&mut group);
-                                }
+                    let fragment_ty = {
```

Comment:
The body of this match arm is the same (I think) as most of the body of the previous arm; could this be factored into a helper function to make it clearer that we do almost the same thing for the two kinds of fragments, and what the differences are in how we treat them?

Reply:
There is now a `collect_fragment` function.